File: misc/piperef.c
/* Copyright © 2016 Bart Massey */

/* Demonstrate pipes and redirecting child stdout. */

#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <errno.h>
#include <string.h>

static char strbuf[1024];

int main(void) {
    int pipefds[2];
    assert(pipe(pipefds) == 0);
    pid_t pid = fork();
    assert(pid >= 0);
    if (pid == 0) {
        assert(close(pipefds[0]) == 0);
        int spid = setsid();
        if (spid == -1) {
            perror("setsid");
            abort();
        }
        assert(dup2(pipefds[1], 1) == 1);
        printf("hello world\n");
        exit(0);
    }
    assert(close(pipefds[1]) == 0);
    FILE *child_stdout = fdopen(pipefds[0], "r");
    assert(child_stdout);
    fgets(strbuf, sizeof(strbuf), child_stdout);
    fprintf(stderr, "got %s", strbuf);
    int wstatus = 0;
    pid_t result = waitpid(pid, &wstatus, 0);
    if (result == -1) {
        perror("waitpid");
        exit(1);
    }
    if (WEXITSTATUS(wstatus) != 0) {
        fprintf(stderr, "child exited with status %d\n", WEXITSTATUS(wstatus));
        exit(1);
    }
    return 0;
}

File: misc/piperef-cargo/piperef.rs
// Copyright © 2016 Bart Massey

// Demonstrate pipes and redirecting child stdout.

extern crate libc;

use std::ffi::CString;
use std::io::Error;

macro_rules! check_cint {
    ($cexp: expr) => (match unsafe { $cexp } {
        -1 => { panic!(Error::last_os_error()); },
        _ => ()
    })
}

macro_rules! try_cint {
    ($cexp: expr) => (match unsafe { $cexp } {
        -1 => { panic!(Error::last_os_error()); },
        r => r
    })
}

macro_rules! check_cptr {
    ($cexp: expr) => ({
        let p = unsafe { $cexp };
        if p.is_null() { panic!(Error::last_os_error()); };
    })
}

macro_rules! try_cptr {
    ($cexp: expr) => ({
        let p = unsafe { $cexp };
        if p.is_null() { panic!(Error::last_os_error()); };
        p
    })
}

fn main() {
    let mut pipefds : [libc::c_int; 2] = [0, 0];
    check_cint!(libc::pipe((&mut pipefds).as_mut_ptr()));
    let pid: libc::pid_t =  try_cint!(libc::fork());
    if pid == 0 {
        check_cint!(libc::close(pipefds[0]));
        check_cint!(libc::setsid());
        check_cint!(libc::dup2(pipefds[1], 1));
        println!("hello world");
        check_cint!(libc::exit(0));
    }
    check_cint!(libc::close(pipefds[1]));
    let r_mode = CString::new("r").unwrap().as_ptr();
    let child_stdout = try_cptr!(libc::fdopen(pipefds[0], r_mode));
    let mut strbuf = [0u8;64];
    check_cptr!(libc::fgets(strbuf.as_mut_ptr() as *mut libc::c_char,
                            strbuf.len() as libc::c_int,
                            child_stdout));
    print!("got {}", std::str::from_utf8(&strbuf).expect("bad string"));
    let mut wstatus = 0;
    check_cint!(libc::waitpid(pid, &mut wstatus as *mut libc::c_int, 0));
    let code = try_cint!(libc::WEXITSTATUS(wstatus));
    if code != 0 {
        panic!("child exited with status {}\n", code);
    }
}
